[mlir][spirv][webgpu] Add lowering of IAddCarry to IAdd

inbelic opened the pull request:
WebGPU does not currently support extended arithmetic, this is an issue when we want to lower from SPIR-V. This commit adds a pattern to transform and emulate spirv.IAddCarry with spirv.IAdd operations

Fixes #65154

inbelic committed:
[mlir][spirv][webgpu] Add lowering of IAddCarry to IAdd

WebGPU does not currently support extended arithmetic, this is an issue
when we want to lower from SPIR-V. This commit adds a pattern to
transform and emulate spirv.IAddCarry with spirv.IAdd operations

Fixes #65154

inbelic commented:
@kuhar @antiagainst I will ping you here for a request to review for a SPIR-V related change. No rush, I understand you are both busy, just before I forget to do so :)

kuhar commented on mlir/lib/Dialect/SPIRV/Transforms/SPIRVWebGPUTransforms.cpp:
```diff
@@ -133,6 +133,48 @@ Value lowerExtendedMultiplication(Operation *mulOp, PatternRewriter &rewriter,
+                         Value rhs) {
+  Location loc = addOp->getLoc();
+  Type argTy = lhs.getType();
+  // Emulate 64-bit addition by splitting each input element of type i32 to
+  // i16 similar to above in lowerExtendedMultiplication. We then expand
+  // to 3 additions:
```
The code assumes that values are 32-bit but it does not actually check that? Is this safe?

inbelic replied:
It was checked before we call the function, but I would support having the type check be more local. As such, I have inlined `lowerCarryAddition` directly to the `matchAndRewrite`.

It should be safe as we can assume that all the integer types have been promoted or demoted to i32 before invoking the transform, sourced from here:
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SPIRV/Transforms/SPIRVWebGPUTransforms.cpp#L150C5-L150C5

kuhar commented on mlir/lib/Dialect/SPIRV/Transforms/SPIRVWebGPUTransforms.cpp:
```diff
@@ -133,6 +133,48 @@ Value lowerExtendedMultiplication(Operation *mulOp, PatternRewriter &rewriter,
+  // Emulate 64-bit addition by splitting each input element of type i32 to
+  // i16 similar to above in lowerExtendedMultiplication. We then expand
+  // to 3 additions:
+  //    - Add two low digits into low resut
+  //    - Add two high digits into high result
+  //    - Add the carry from low result to high result
```
Could we calculate the carry result by comparing the sum to one of the operands?
> Member 1 of the result gets the high-order (carry) bit of the result of the addition. That is, it gets the value 1 if the addition overflowed the component width, and 0 otherwise.

I think it's something like:
```
carry := (a + b) < a ? 1 : 0
```

Where `<` is unsigned less than.

This way we wouldn't have to split the values at all.

inbelic replied:
Nice, that feels much better. From now reading the linked spec a second time, it is clear to me that we can expect the overflow behaviour to be how we would want. I was not diligent enough when reading the "avoid overflow and underflow" statement the first time, whoops.

https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_arithmetic_instructions

kuhar reviewed (changes requested):
Thanks for taking a stab at this! I think we can simplify the code, but do let me know if I missed something there.

inbelic committed:
review comments:

- inline the lowerAddCarry function to make it clearer that we check
  that each integer is of i32

- switch the computation of the carry to reduce instructions and
  simplify

kuhar commented on mlir/lib/Dialect/SPIRV/Transforms/SPIRVWebGPUTransforms.cpp:
```diff
@@ -220,13 +178,26 @@ struct ExpandAddCarryPattern final : OpRewritePattern<IAddCarryOp> {
+    Value zero =
+        rewriter.create<ConstantOp>(loc, argTy, getScalarOrSplatAttr(argTy, 0));
+
+    // Emulate 64-bit unsigned addition by allowing our addition to overflow,
```
This doesn't emulate 64-bit addition, it performs addition and returns an extra carry result

kuhar replied:
Maybe something like: `// Calculate the carry by checking if the addition resulted in an overflow.`

kuhar reviewed (commented):
Nice, thanks for the changes.

Could you also add a runtime integration test under https://github.com/llvm/llvm-project/tree/main/mlir/test/mlir-vulkan-runner to make sure we are getting the expected results?

inbelic committed:
review comments

- clarify description comment
- add integration test for the vulkan runner

inbelic commented:
Rebased and force pushed to hopefully remedy the code_formatter error.
I added an integration test, but it is only done by eye. I am not sure how one can test it locally? Is there documentation for the possible ways to test with the `mlir-vulkan-runner`?

kuhar commented:
> Rebased and force pushed to hopefully remedy the code_formatter error. I added an integration test, but it is only done by eye. I am not sure how one can test it locally? Is there documentation for the possible ways to test with the `mlir-vulkan-runner`?

You need to set these two in cmake:
```
      -DMLIR_INCLUDE_INTEGRATION_TESTS=1 \
      -DMLIR_ENABLE_VULKAN_RUNNER=1 \
```

kuhar commented:
The runner test fails with:
```
******************** TEST 'MLIR :: mlir-vulkan-runner/iaddcarry_extended.mlir' FAILED ********************
Exit Code: 2

Command Output (stdout):
--
# RUN: at line 4
mlir-vulkan-runner /home/jakub/llvm/llvm-project/mlir/test/mlir-vulkan-runner/iaddcarry_extended.mlir   --shared-libs=/home/jakub/llvm/build/relass/lib/libvulkan-runtime-wrappers.so,/home/jakub/llvm/build/relass/lib/libmlir_runner_utils.so   --entry-point-result=void | /home/jakub/llvm/build/relass/bin/FileCheck /home/jakub/llvm/llvm-project/mlir/test/mlir-vulkan-runner/iaddcarry_extended.mlir
# executed command: mlir-vulkan-runner /home/jakub/llvm/llvm-project/mlir/test/mlir-vulkan-runner/iaddcarry_extended.mlir --shared-libs=/home/jakub/llvm/build/relass/lib/libvulkan-runtime-wrappers.so,/home/jakub/llvm/build/relass/lib/libmlir_runner_utils.so --entry-point-result=void
# .---command stderr------------
# | loc("/home/jakub/llvm/llvm-project/mlir/test/mlir-vulkan-runner/iaddcarry_extended.mlir":25:59): error: expected ','
# | could not parse the input IR
# `-----------------------------
# error: command failed with exit status: 1
# executed command: /home/jakub/llvm/build/relass/bin/FileCheck /home/jakub/llvm/llvm-project/mlir/test/mlir-vulkan-runner/iaddcarry_extended.mlir
# .---command stderr------------
# | FileCheck error: '<stdin>' is empty.
# | FileCheck command line:  /home/jakub/llvm/build/relass/bin/FileCheck /home/jakub/llvm/llvm-project/mlir/test/mlir-vulkan-runner/iaddcarry_extended.mlir
# `-----------------------------
# error: command failed with exit status: 2
```

kuhar commented:
@inbelic I made the test type check and compile but it doesn't give the expected output: https://github.com/kuhar/llvm-project/commit/d20f576f89b6595549126fde8311a0c4bf93e104

I changed the i8 memrefs to i32 because the type of both iadd_carry results will be the same on the spir-v side anyway, and it doesn't require spir-v capabilities for i8 types. This extends out test coverage to vulkan implementations that don't expose these, e.g., swiftshader.

inbelic committed:
fix integration test

- use int32 to reduce required capabilities of testing environments
- updated test values with the expected behaviour

inbelic commented:
@kuhar Awesome thanks. Have got the runner test locally and also get the wrong behaviour when using the negative minimum integers for the input.  I believe this is because the `Signedness` is set for the integer type. While `arith.addui_extended` and `spirv.IAddCarry` will interpret the bits as unsigned, `spirv.IAddOp` does not and that is why it does not provide our desired behaviour to overflow when we add 1 to it. All of this to say, that if you set `-2^31-1` to `2^32-1` instead, then we have the expected behaviour.

However, imo it should be that the negative number is interpreted as an unsigned integer and then we unfortunately will need to use a different solution.

kuhar commented:
@inbelic Signed and unsigned integer addition is the same operation, that's why there's a single `addi` instead of  `addui`/`addsi`. When we embed constants in the IR and print it back the interpretation is signed and we see negative numbers, but it doesn't matter -- it's the same bitpattern as for the unsigned result.

Therefore, `addui_extended`/`IAddCarry` both implement signless addition, and the carry bit corresponds to the unsigned overflow bit.

kuhar reviewed (approved):
LGTM. I ran the integration tests on my machine and everything passed. Thanks!

kuhar commented:
@inbelic I went ahead and merged the code
